phpseclib/File/X509.php

```diff
@@ -1137,7 +1138,11 @@ public function validateSignature($caonly = true)
                                 $subjectKeyID = $this->getExtension('id-ce-subjectKeyIdentifier', $ca);
                                 switch (true) {
                                     case !is_array($authorityKey):
+                                    case !$subjectKeyID:
```

Comment:
What if the signed cert has authority key identifier and no authority cert serial number but the candidate CA cert doesn't have a subject key identifier? I can write a test to be sure but it looks like the old code would have continued looping whereas it looks like this could would think such a cert was the CA cert when it probably ought not be.

Reply:
That case will never happen because we are checking first:
if ca.subject == cert.issuer (lines 1135 and 1136)
so if ca.subject and cert.issuer are the same we just have to perform additional checks on fields like serial if are set.

Reply:
> so if ca.subject and cert.issuer are the same we just have to perform additional checks on fields like serial if are set.

That's what I'm saying. Your code isn't doing that. And lines 1135 and 1136 aren't doing the additional checks you speak of.

Here's an example.

The signing cert:

```
-----BEGIN CERTIFICATE-----
MIIB4zCCAUygAwIBAgIUSyCTib61O8k8TM5JA5n3zDYDFs8wDQYJKoZIhvcNAQEF
BQAwHDEaMBgGA1UECgwRcGhwc2VjbGliIGRlbW8gQ0EwIBcNMTcxMjEwMTYxMDIw
WhgPOTk5OTEyMzEyMzU5NTlaMBwxGjAYBgNVBAoMEXBocHNlY2xpYiBkZW1vIENB
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCn5XdGJUd9PY9zpksbJ2kD57PL
nQ3ISK+qbeSe2H+Uj0AEDnm611tvXlr9a7zBPNwZ2ali9bczirQkxtu/sftSHCGv
LL6sz1vdYP0Un792o5h+PkbQLnDe1lf/W8Q87GY9cRxUQVVfA7QUfgsl4iWgi7tv
qz3dRmxzBpjpgWvp/QIDAQABoyAwHjALBgNVHQ8EBAMCAQYwDwYDVR0TAQH/BAUw
AwEB/zANBgkqhkiG9w0BAQUFAAOBgQBQJpF/uakVjNIOz4Ve+duZYeeghpVxVNVK
t4OFi8DQzAsdFBpDDg/ZdAoXYOI/VHwNR+jKS2bPLmvNR2dP+OuZVdSTYu4/dZtI
Tp9ju17bym9oyWjX8CApa32+4N4WtcYLaVHOgyGA9f1VRHBtLTc8xcuHnJg6v3m9
TYVkR8c3IQ==
-----END CERTIFICATE-----
```

The signed cert:

```
-----BEGIN CERTIFICATE-----
MIIB5TCCAU6gAwIBAgIUVCrbYZRButS1ka5fyUe2iOzFAmwwDQYJKoZIhvcNAQEF
BQAwHDEaMBgGA1UECgwRcGhwc2VjbGliIGRlbW8gQ0EwIBcNMTcxMjEwMTYxMDIw
WhgPOTk5OTEyMzEyMzU5NTlaMB4xHDAaBgNVBAoME3BocHNlY2xpYiBkZW1vIGNl
cnQwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBANkmWIAsKkL1Bv3PiFA65IRQ
EOSc65rui70M5tvrDpka1jl6QgvGY/jI/6D6h8sZBxyYyncgWaCA4dVv9dmdZm8Z
JPXACUfHTJuk7ApPq2oYjyheyQkSLp8kYP47cidbNfWk3RkTRqwleh3ByG4e1Qxh
sfdf5owA1yuW5JSygppvAgMBAAGjIDAeMA4GA1UdIwQHMAWAA3p6ejAMBgNVHQ4E
BQQDenp6MA0GCSqGSIb3DQEBBQUAA4GBACMLNye3bEzTnHJAroAiV8Rw9rL6QG+x
wrFDBq173MXTAnKar/TujUxfmOzNZCxjPHh6oKRosOWZuI5v8/0wG9I6irYwLXOw
kXMvXT91/m5Vk7rLCqQ+zsFncxg12leHRgglWh1wsJOhPI/B0mYgCA04RNHsm+iC
/XjRBlAt1kMb
-----END CERTIFICATE-----
```

phpseclib doesn't recognize this cert as being valid (and I'm not sure it ought to) but in your changes it is recognized as valid. The signing cert doesn't have a subject key identifier while the signed cert does have an authority key identifier (and a subject key identifier).

That said, to be fair, OpenSSL seems to think the cert is valid:

```
openssl verify -verbose -CAfile signer.pem signed.pem
```

That said, idk that I agree with OpenSSL's verdict. In light of this I've posed the question on crypto.stackexchange.com:

https://crypto.stackexchange.com/q/53856/4520

Reply:
So I've re-evaluated things after the crypto.stackexchange.com and I'll go ahead and merge this. I'll update it based on my comment at https://github.com/phpseclib/phpseclib/pull/1213#discussion_r153625564 , I'll backport to 1.0 and I'll see if this should be added for CRL's (`tbsCertList`).

Reply:
Great! Thanks a lot!